pkg/apb/watch_pod.go

```diff
@@ -32,45 +34,63 @@ var (
 	ErrorActionNotFound = fmt.Errorf("action not found")
 )
 
-func watchPod(podName string, namespace string) error {
+func watchPod(podName string, namespace string, podClient v1.PodInterface, statusUpdates chan<- JobState) error {
 	log.Debugf(
 		"Watching pod [ %s ] in namespace [ %s ] for completion",
 		podName,
 		namespace,
 	)
 
-	k8scli, err := clients.Kubernetes()
```

Comment:
Why don't we get the `v1.PodInterface`  in this method?

I think that is better because it now ever caller has to get the client for the only purpose of passing into a function. I think this encapsulates that behaviour and if it errors we bubble up that error.

If the only reason we are doing it is for testing then I think we need to do a better job of mocking out the core componets rather then passing values into functions. I think that we got bite by doing that before.

Reply:
If we want unit tests for a method like this, then at some point we will need to use an interface to allow us to mock the interactions with external dependencies. 
Having a package level function used internally as part of this method, means we cannot mock it out. 
Really this is just a simple form of dependency injection. The package level functions can make this tricky.  I would prefer to have a unit test for the logic here, than not have one?  I took a similar approach to the provision function here
https://github.com/maleck13/ansible-service-broker/blob/4f0301dac3dae580cb5f393baafaa70a2d859073/pkg/broker/broker.go#L377
Except it is passed in via the constructor. It would be a bigger refactor to 
In this case to have the dependency passed as part of a constructor as it would likely mean changes all the way up the call chain.

On that point though, longer term it may be worth discussion to change some of the package functions to be struct methods and make use of the ```kubernetes.Interface``` in place of the ```*client.ClientSet``` concrete type. 
If we use the ```kubernetes.Interface``` and refactor to allow it to be injected as part of the main method via constructors and dependency injection, it will allow us to have a very powerful set of unit tests as we can control the behavior of the tests by mocking out just the client as generally this is what underpins all of the external communications. This becomes even more true if we move to CRDs instead of a separate etcd.

I can create a follow on issue for this.